File: java/L700/LeetCode0714.java
package L700;

public class LeetCode0714 {
    public int maxProfit(int[] prices, int fee) {
        if (prices == null || prices.length <= 1) {
            return 0;
        }
        int hold = -prices[0];
        int not_hold = 0;
        for (int i = 1;i < prices.length;++i) {
            int h = hold, n = not_hold;
            hold = Math.max(h, n + prices[i]);
            not_hold = Math.max(n, h + prices[i] - 2);
        }
        return Math.max(hold, not_hold);
    }
}

File: cpp/L700/LeetCode0714.cc
#include <limits>
#include <vector>
using std::vector;

class Solution {
public:
    int maxProfit(vector<int>& prices, int fee) {
        int hold{std::numeric_limits<int>::min() + 10000000}, unhold{0};
        for (int price : prices) {
            hold = std::max(hold, unhold - price);
            unhold = std::max(unhold, hold + price - fee);
        }
        return std::max(hold, unhold);
    }
};